### PR TITLE
Gang Machinegun Mouse Tracking + Projectile Source Fix

### PR DESCRIPTION
**Pull request by silicons**
1. Moves Gang Machineguns to its own file. It's not a goddamn portable turret and it's not even something that uses power!
2. Gang Machineguns now track the mouse location of the first buckled person (For those who don't code, that means the person using it)
3. Gang Machineguns will automatically aim and turn towards whatever that person is pointing at.. However, to make it target something specifically instead of a direction, you still have to hold your mouse 
down while firing.
4. Gang machineguns will no longer cause the user to shit out bullets if they unbuckle during firing :^)

robustin_is_bad_at_coding

# SOME REFACTORS TO PIXEL PROJECTILES ARE INCLUDED...

:cl:
experimental: Gang turrets now follow the mouse of the person using them! Yay!
/:cl:

### CODE REVIEW

**Comment by GunHog** [comment]
**DEATH BLOSSOM MODE ENGAGED!**

**Comment by Robustin** [comment]
Good fixes but permanent autotracking is not an improvement. I need to be able to examine stuff without necessarily wanting to point my gun at/shoot at it. 

But holding down mouse isn't the most intuitive aiming choice either so meh.

**Comment by kevinz000** [comment]
Why would you not turn your gun around to examine something if it's behind you though

**Comment by ChangelingRain** [comment]
the turning is so jerky aaaaaaaaa

**Comment by Thunder12345** [comment]
Can you make the fire sound a var by any chance?

**Comment by kevinz000** [comment]
@ChangelingRain I'm too burned out to make it smooth 360 degrees right now maybe later :3

**Comment by kevinz000** [comment]
@GunHog that'd be a balance thing bug @Robustin

**Comment by evsey9** [comment]
GIVE IT A LASER BEAM AIM

**Comment by GunHog** [comment]
Just do it, it is fine, Kev. It is an advantage in that the user can see where he aims, but the disadvantage is that everyone else can see it as well!

File: code/game/objects/structures/manned_turret.dm [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+		user.client.change_view(view_range)
+	START_PROCESSING(SSfastprocess, src)
+
+/obj/machinery/manned_turret/process()
+	if(!LAZYLEN(buckled_mobs))
+		STOP_PROCESSING(SSfastprocess, src)
+		return
+	var/mob/living/controller = buckled_mobs[1]
+	if(controller.client)
+		var/client/C = controller.client
```
Comment (Cyberboss) [inline code/game/objects/structures/manned_turret.dm]: Swap these 2 lines and do if(C)

File: code/game/objects/structures/manned_turret.dm [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+	w_class = WEIGHT_CLASS_HUGE
+	flags = ABSTRACT | NODROP | NOBLUDGEON
+	resistance_flags = FIRE_PROOF | UNACIDABLE | ACID_PROOF
+	var/obj/machinery/manned_turret/turret
+
+/obj/item/gun_control/New(obj/machinery/manned_turret/MT)
+	if(MT)
+		turret = MT
+	else
+		qdel(src)
```
Comment (Cyberboss) [inline code/game/objects/structures/manned_turret.dm]: Replace this whole proc with:
```
/obj/item/gun_control/Initialize()
    . = ..()
    turret = loc
    if(!istype(turret))
        return INITIALIZE_HINT_QDEL
```

File: code/game/objects/structures/manned_turret.dm [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+			user.pixel_y = -4
+
+/obj/item/gun_control
+	name = "turret controls"
+	icon = 'icons/obj/weapons.dmi'
+	icon_state = "offhand"
+	w_class = WEIGHT_CLASS_HUGE
+	flags = ABSTRACT | NODROP | NOBLUDGEON
+	resistance_flags = FIRE_PROOF | UNACIDABLE | ACID_PROOF
+	var/obj/machinery/manned_turret/turret
```
Comment (Cyberboss) [inline code/game/objects/structures/manned_turret.dm]: Add a Destroy() to null this

File: code/game/objects/structures/manned_turret.dm [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+/obj/machinery/manned_turret/user_buckle_mob(mob/living/M, mob/living/carbon/user)
+	if(user.incapacitated() || !istype(user))
+		return
+	M.forceMove(get_turf(src))
+	..()
+	for(var/V in M.held_items)
+		var/obj/item/I = V
+		if(istype(I))
+			if(M.dropItemToGround(I))
+				var/obj/item/gun_control/TC = new /obj/item/gun_control(src)
```
Comment (Cyberboss) [inline code/game/objects/structures/manned_turret.dm]: Just `= new(src)`

File: code/game/objects/structures/manned_turret.dm [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+				M.put_in_hands(TC)
+		else	//Entries in the list should only ever be items or null, so if it's not an item, we can assume it's an empty hand
+			var/obj/item/gun_control/TC = new /obj/item/gun_control(src)
```
Comment (Cyberboss) [inline code/game/objects/structures/manned_turret.dm]: Just `= new(src)`

File: code/game/objects/structures/manned_turret.dm [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+/obj/machinery/manned_turret/user_buckle_mob(mob/living/M, mob/living/carbon/user)
+	if(user.incapacitated() || !istype(user))
+		return
+	M.forceMove(get_turf(src))
+	..()
+	for(var/V in M.held_items)
+		var/obj/item/I = V
+		if(istype(I))
+			if(M.dropItemToGround(I))
```
Comment (Cyberboss) [inline code/game/objects/structures/manned_turret.dm]: They should just be unbucklable if they're holding something

File: code/game/objects/structures/manned_turret.dm [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+	playsound(src,'sound/mecha/mechmove01.ogg', 50, 1)
+	anchored = TRUE
+	if(user.client)
+		user.client.change_view(view_range)
+	START_PROCESSING(SSfastprocess, src)
+
+/obj/machinery/manned_turret/process()
+	if(!LAZYLEN(buckled_mobs))
+		STOP_PROCESSING(SSfastprocess, src)
+		return
```
Comment (Cyberboss) [inline code/game/objects/structures/manned_turret.dm]: these 2 lines -> return PROCESS_KILL

File: code/game/objects/structures/manned_turret.dm [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+		volley(user)
+
+/obj/machinery/manned_turret/proc/volley(mob/user)
+	target_turf = get_turf(target)
+	for(var/i in 1 to number_of_shots)
+		addtimer(CALLBACK(src, /obj/machinery/manned_turret/.proc/fire_helper, user), i*rate_of_fire)
+
+/obj/machinery/manned_turret/proc/fire_helper(mob/user)
+	if(!src)
+		return
```
Comment (Cyberboss) [inline code/game/objects/structures/manned_turret.dm]: Remove these lines

File: code/game/objects/structures/manned_turret.dm [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+	process()						//REFRESH MOUSE TRACKING!!
```
Comment (Cyberboss) [inline code/game/objects/structures/manned_turret.dm]: Seperate what you need in process into a unique proc. Don't call it directly

File: code/game/objects/structures/manned_turret.dm [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+		if(WEST)
+			layer = ABOVE_MOB_LAYER
+			user.pixel_x = 14
+			user.pixel_y = 0
+		if(NORTHWEST)
+			layer = BELOW_MOB_LAYER
+			user.pixel_x = 8
+			user.pixel_y = -4
+
+/obj/item/gun_control
```
Comment (Cyberboss) [inline code/game/objects/structures/manned_turret.dm]: Move all these definitions below the manned_turret stuff

File: code/game/objects/structures/manned_turret.dm [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+	if(!src)
+		return
+	process()						//REFRESH MOUSE TRACKING!!
+	var/turf/targets_from = get_turf(src)
+	if(QDELETED(target))
+		target = target_turf
+	var/obj/item/projectile/P = new projectile_type(targets_from)
+	P.current = targets_from
+	P.starting = targets_from
+	loc = targets_from
```
Comment (Cyberboss) [inline code/game/objects/structures/manned_turret.dm]: forceMove

File: code/modules/projectiles/projectile.dm [hunk 12]
```diff
@@ -282,43 +282,51 @@
...
-			//Calculate the "resolution" of screen based on client's view and world's icon size. This will work if the user can view more tiles than average.
-			var/screenview = (user.client.view * 2 + 1) * world.icon_size //Refer to http://www.byond.com/docs/ref/info.html#/client/var/view for mad maths
-
-			var/ox = round(screenview/2) - user.client.pixel_x //"origin" x
-			var/oy = round(screenview/2) - user.client.pixel_y //"origin" y
-			// to_chat(world, "Pixel position: [x] [y]")
-			var/angle = Atan2(y - oy, x - ox)
-			// to_chat(world, "Angle: [angle]")
-			src.Angle = angle
+	loc = get_turf(user)
```
Comment (Cyberboss) [inline code/modules/projectiles/projectile.dm]: forceMove

File: code/game/objects/structures/manned_turret.dm [hunk 13]
```diff
@@ -0,0 +1,199 @@
...
+	obj_integrity = 100
+	buckle_lying = 0
+	layer = ABOVE_MOB_LAYER
+	var/view_range = 10
+	var/cooldown = 0
+	var/projectile_type = /obj/item/projectile/bullet/weakbullet3
+	var/rate_of_fire = 1
+	var/number_of_shots = 40
+	var/cooldown_duration = 90
+	var/atom/target
```
Comment (Cyberboss) [inline code/game/objects/structures/manned_turret.dm]: Null this in Destroy()

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Comment by evsey9** [comment]
and make all vigiliantes on screen turn into facehuggers for the person on the mounted machinegun turret

File: code/game/objects/structures/manned_turret.dm (hunk 3 above)
Comment (kevinz000) [inline code/game/objects/structures/manned_turret.dm]: done

File: code/game/objects/structures/manned_turret.dm (hunk 6 above)
Comment (kevinz000) [inline code/game/objects/structures/manned_turret.dm]: @Robustin

File: code/game/objects/structures/manned_turret.dm (hunk 13 above)
Comment (kevinz000) [inline code/game/objects/structures/manned_turret.dm]: done

**Pull request merged**